### ISSUE TITLE
"Load More" functionality doesn't work when passing in searchState to component using InfiniteHits 

### ISSUE THREAD

**Issue by tedlin182** [algolia/instantsearch.js#1758]
**Setup:**
- An "All Results" component that displays limited # of results for various categories
- A "Show More" component using `InfiniteHits` widget that displays certain # of results but ideally would add more results after clicking "Load More"

**Desired Behavior**
- "All Results" component shows first as you type in a query and each category section of results has a "Show more" link to bring in the "Show More" component.
- We want to preserve the current query and show the same results as displayed on "All Results" component and as you click "Load More", more results would get added on.

If I don't pass in the current `searchState` prop to the "Show more" results component, the Load More works just fine, but the current query isn't present.

When I do pass it in, the current query is preserved and the same results show, but the Load More stops working (is not disabled).

In the end, passing `searchState` that has a non-empty `query` to the `InstantSearch` wrapping `InfiniteHits`, seems to break the Load More functionality and it's not very obvious why, at least to me. Unsure if I'm missing something or if there's a better way of doing this, but let me know if anyone has any ideas.

**Update:**

After some debugging, I located part of thethe issue which was within `connectInfiniteHits` in the `refine` function. Basically the `searchState` I pass in with the current query `{query: "some text"}`  overwrites the `searchState` that is generated somewhere else and passed into `refine`. This loses the `page` key-value from the internally generated `searchState` and thus continues to reset to page 1 since `page` isn't found in `searchState`.

Need to do some more digging as to where the `searchState` for `refine` is getting passed in from, but wherever it is, it should just merge in the `searchState` prop that we pass into `InstantSearch` instead of having it completely replaced.

**Comment by mthuret** [comment]
Hi @tedlin182, thanks for reaching us with your use-case. 

You're right when passing the `searchState` props to an `<InstantSearch />`instance, you need to provide the entire state, we don't perform any kind of merge. But if you are listening to the `onSearchStateChange` hook, you are able to retrieved the entire state. Are you using it?

Also to have a better understanding of what you are trying to achieve, maybe you can provide us a gif showing what you have today in term of behavior and some simplify code showing how you use the library? (if you can link a repo it would even be easier). 

I also have some questions:
What behavior requires you to pass a `searchState` to an `<InstantSearch/>` instance? 
Which part of the `react-instantsearch` library are you using to build your `All Results` component? Which widget or connector?
When you talk about the "Load More" component using the `InfiniteHits` widget, I assume that you only use the "load more" button and hide the rest?

**Comment by tedlin182** [comment]
Hey @mthuret !

Thanks for your help!

So I forgot I was using a custom input text field component instead of your `SearchBox` widget, so we were manually setting the `searchState.query` down thus it was overwriting it. I replaced it back with your `SearchBox` and the load more seems to work.

However, now I have another issue where after I click show more, load the "show more" view and then click back to the "all results" view, it clears out the `SearchBox` query, triggers the `onSearchStateChange` and resets the `searchState` to `{page: 0}`.

The `ShowMoreResults` component uses `InfiniteHits` and `AllResults` uses `Hits`. The only thing I am doing to change the views is changing a state value in my parent component from `true` to `false`.

**Comment by mthuret** [comment]
When a component is unmounted we remove its state from `searchState`, so if you need to keep track of the `page` cursor, you might want to save it somehow and add it to the `searchState` when mounting your `<ShowMoreResults/>` component.

I'm curious though about the behavior you put in place. If I understand correctly your `<AllResults/>` component is the combination of all your <Category/> components using each one its own `<InstantSearch/>` with its own index. But what is the configuration of the `<InstantSearch/>` used by your `<ShowMoreResults/>` component? Is there only one global button, or one per category?

**Comment by tedlin182** [comment]
Hey @mthuret 

Sorry I forgot to post a gif of the behavior we want. I had to run to catch a flight, but it's below for you to view.

The `SiteSearch` component is the main wrapper component that contains both `AllResults` and `ShowMoreResults`. We toggle between the `AllResults` and `ShowMoreResults` views via a state change (`viewShowMore') in `SiteSearch`.

**`AllResults` **
- This contains each category's results, each having their own "Show more" link that brings in the `ShowMoreResults` component/view.
- Each category results component have their own `<InstantSearch/>` with their own index, `<Hits/>`, and `<VirtualSearchBox/>`.
- I pass in the `searchState` and `onSearchStateChange` method down from `SiteSearch` to each of these category results components.
- Also using `createConnector` to hide results when not matches are found.

**`ShowMoreResults`**
- This contains a specific category's results that match the current query in the `<SearchBox/>`
- It has its own `<InstantSearch/>` with their own index, `<InfiniteHits/>`, and `<VirtualSearchBox/>`
- We limit the initial display to show 5 results and want the user to be able to load more matching results for that specific category when they click "Load More".
- There's also a "Back to all results" link that just toggles the state `viewShowMore` in `SiteSearch` that unmounts the `ShowMoreResults` component/view and brings back in the `AllResults` view.
- Also using `createConnector` to hide results when not matches are found.
- **Note:** We don't unmount `AllResults` ever, just `ShowMoreResults`

I tested out NOT unmounting `ShowMoreResults` and it seemed to resolve that issue regarding the reset `searchState` but then I ran into other issues.

Here is a video of the behavior we're trying to achieve and at the very end you'll notice the results dropdown closing automatically when we hit "Back to results". That is the part where `query` isn't returned back in `searchState` after we unmount `ShowMoreResults`.

<http://recordit.co/jMTPlwepYJ>

Let me know if you want any other info but hopefully this helps.

Thanks!

**Comment by mthuret** [comment]
@tedlin182 Thanks for the gif, it's definitely clearer :) 

So yes, either you manage not to unmount the <ShowMoreResults/> component, or your <SitreSearch/> component should keep track of the page cursor (this way, you can control when to reset it, or not).

I did have a suggestion though, that may simplify your code. Why not only use the `connectInfiniteHits`connector to achieve the same behavior. 

* By default, when typing a query, each category will be displayed with 5 results and a show more button. 
* When clicking on a show more button, you actually hide every other categories and just continue to display hits from the wanted one.

Do you think this could work for you?

**Comment by tedlin182** [comment]
Hey @mthuret !

Are you suggesting to use `connectInfiniteHits` in just the `ShowMoreResults` component instead of `createConnector`? I thought about seeing if I could use the `connectInfiniteHits` connector but I wasn't completely sure as to how to actually implement it. Not sure if you had an example for this but that would be helpful. Would I would wrap the `ShowMoreResults` component in `connectInfiniteHits` when I mount it in `SiteSearch`?

Yeah when I left the `ShowMoreResults` component mounted and tried to maintain the same behavior, clicking "show more" would initially show the correct results, but if I clicked "back to results" and then "show more" in a different category, it wouldn't update the results, so it would display the previous category results in the "show more" view. Not until I went back and clicked "show more" would it update so it seemed as if there was a lag maybe.

**Comment by mthuret** [comment]
My suggestion is not to use the `<Hits/>` widgets anymore, and basically merge the AllResults/ShowMoreResults component in one. I believe you will still need to have the createConnector to handle the NoResults feature. 

here is an implementation example for the connectInfiniteHits: https://github.com/algolia/instantsearch.js/blob/v2/docgen/src/examples/e-commerce-infinite/App.js
([live](https://community.algolia.com/instantsearch.js/react/examples/e-commerce-infinite/)

You would have a structure like:

Site search (Handle the hide/show of the category when needed and the animation)
-- Category 1 (connectInfiniteHits)
-- Category 2 (connectInfiniteHits)
-- Category 3 (connectInfiniteHits)

I suggest `connectInfiniteHits` but maybe only the `<InfiniteHits/>` can be enough. The behavior of this widget is to display N hits and display a `show more` button to display more hits. Is this clearer?

**Comment by tedlin182** [comment]
Thanks for the info and examples, definitely helpful. 

Just so I'm understanding your suggestion:
- Basically each category results component will be wrapped in `connectInfiniteHits`. 
- After a user types in a query in the search field and clicks "show more results" for a specific category, it will hide the other category results components via a state change in `SiteSearch` (`viewShowMore`) and only show the targeted category results component with the "Load More" button.
- Clicking on the "Back to all results", will re-show all of the all category results components, hide the "Load More" button, and all `searchState` values will persist through all of these views and interactions.

A few questions:
* Since the `page` key-value affect which/how many results show in each category, will this get reset back to `0` if a user clicks "Load more" at least once in the "show more" view and then clicks "back to results"? We want to always show the first 5 results in each category in the "All Results" view.
* How can I hide the "Load more" button in the category results component while in the "All Results" view and then show it while in the "ShowMoreResults" view? Ideally, I could do this via passing in the "viewShowMore" state value from `SiteSearch` down to these category results components via a prop and then choose to show or hide it.
* In your proposal, will I still be able to maintain the exact same UI behavior as in the GIF I posted earlier? In order to get the sliding affect, I would have to position that "show more" category results component offscreen on "viewShowMore" state change and then slide it in, but I'd have to test to see if there will be any jumpiness in the UI.

Let me about the above and maybe I'll test out your suggestions to see if I can get this working.

**Comment by mthuret** [comment]
Hey @tedlin182 

1. The "back to results" link will have to reset the `page` to 0 for the given category. If you don't want to keep the query, then you should also clean up this one. 
2. You can keep a state inside your `<SiteSearch/>` component
3. This is the part where I don't know :)

Also, please let me know if keeping in the state the page cursor works better than not unmouting your <ShowMoreResults/> component.

**Comment by tedlin182** [comment]
@mthuret 

Thanks! Could you clarify what you mean by keeping tracking of the page cursor? Are you talking about the cursor focus in the search field?

**Comment by tedlin182** [comment]
@mthuret 

For the time being, I actually am going to keep my current component structure and created a temporary object container in `SiteSearch` to remember the `query` from `searchState` when `ShowMoreResults` unmounts and to merge in that query only if it exists. I then set it back to `null` afterwards as I use it as an indicator that I'm coming from the `ShowMoreResults` view.

Seems to work and I'll keep investigating and testing out some of the suggestions you made above to see if that could replace what I have now.

Thanks!

**Comment by mthuret** [comment]
Good to hear that you manage to have something working! That's nice @tedlin182 👍

**Comment by mthuret** [comment]
I'll close this issue, don't hesitate to reopen it if something is still wrong :)

**Comment by Alexis-bf** [comment]
result object work for me, in angular i made this *ngIf="hits.length !== results.nbHits"
<button
              *ngIf="hits.length !== results.nbHits"
              class="btn-primary col-12 col-lg-4 col-sm-6 btn-load-me"
              (click)="refine()"
            >
              Load more events
            </button>
